### ISSUE TITLE
OTA update fails

### ISSUE THREAD

**Issue by AieatAssam** [SeeedJP/ReButton#14]
I have ReButton version 1.0 and can see that firmware is currently on 1.4 and has added support for custom messages. Following update instructions on https://github.com/SeeedJP/ReButton/blob/master/OTA.md I wasn't getting anywhere. So I connected my trusty USB-UART to the debu port to try to see what is going on (baud rate should be 115200, by the way - doesn't seem to be documented, so I had to dig into arduino board source for this).

Results are as follows:
```
packageURI = https://seeedkk.blob.core.windows.net/rebuttonfirm/ReButtonApp.1.4.bin
packageCRC = 0xdaeb
packageSize = 680740
Connect WiFi...
Time from pool.ntp.org, now is (UTC): Thu Jan  1 00:00:06 1970

Connected WiFi.
Download firmware...
>>> Failed to send telemetry data: Http fault.
Downloaded firmware.
inPackageCRC = 0xdaeb
inPackageSize = 680740
dlPackageCRC = 0x02
dlPackageSize = -1

```

Looking into the board files and OTAFirmwareUpdate.cpp, it seems that size -1 occurs when failing to write to flash. Just in case, to eliminate SSL being the issue, I created Azure blob container without secure access enabled and tried OTA update without SSL. Results are the same. 

Is there a step I am missing to make flash writeable for OTA updates? Or is my hardware faulty? I only received it yesterday and ordered direct from Seeed, so was hoping it would have up to date firmware to begin with (no such luck).
Are there any other ways of uploading firmware? i.e. do you have bootloader mode that can take it via UART? Although doubt it would do much if flash itself is non-writeable (line setting body length to -1 is the relevant one):
```C++
static void getFwCallback(const char *at, size_t length)
{
    if (at == NULL || length == 0 || body_length < 0)
    {
        return;
    }
    // Write to external flash
    if (MicoFlashWrite((mico_partition_t)MICO_PARTITION_OTA_TEMP, &flashIdx, (uint8_t *)at, length) == kNoErr)
    {
        CRC16_Update(&contexCRC16, at, length);
        body_length += length;
    }
    else
    {
        body_length = -1;
    }
}

int OTADownloadFirmware(const char *url, uint16_t * crc16Checksum, const char* ssl_ca_pem)
{
    uint16_t checkSum = 0;
    flashIdx = 0;
    body_length = 0;
    CRC16_Init(&contexCRC16);

    HTTPClient client = ssl_ca_pem ? HTTPClient(ssl_ca_pem, HTTP_GET, url, getFwCallback) : HTTPClient(HTTP_GET, url, getFwCallback);
    const Http_Response *response = client.send(NULL, 0);
    if (response->status_code != 200)
    {
        // Download failed
        return -1;
    }
    else if (body_length < 0)
    {
        // External flash accessing issue
        return -2;
    }
    else if (body_length == 0)
    {
        // Empty
        return 0;
    }

    // Finalize the CRC16 value
    CRC16_Final(&contexCRC16, &checkSum);

    if (crc16Checksum)
    {
        *crc16Checksum = checkSum;
    }

    return body_length;
}
```

**Comment by ytsuboi** [comment]
Hi @AieatAssam ,
As on the OTA documentation Step 1&2, you need to store the binary on Azure blob container. We are not sure why but the boot loader we are using ( provided as binary ) only supports downloading from blob container.
The debug log you attached pointed our github URL.

**Comment by AieatAssam** [comment]
Actually, my log points to your blob at seeedkk.blob.core.windows.net

But I have also tried with my own blob storage accounts - one with and one without SSL and both times got the same result. Exactly the same size -1.

**Comment by matsujirushi** [comment]
Hi @AieatAssam ,

> Time from pool.ntp.org, now is (UTC): Thu Jan  1 00:00:06 1970

I think your ReButton failed time sync.
Could you change time server in Wi-Fi page?

**v1.0 -> v1.4 log:**  
```
packageURI = https://seeedkk.blob.core.windows.net/rebuttonfirm/ReButtonApp.1.4.bin
packageCRC = 0xdaeb
packageSize = 680740
Connect WiFi...
Time from pool.ntp.org, now is (UTC): Wed Sep  4 00:34:45 2019

Connected WiFi.
Download firmware...
Downloaded firmware.
inPackageCRC = 0xdaeb
inPackageSize = 680740
dlPackageCRC = 0xdaeb
dlPackageSize = 680740
Switch firmware...
Switched firmware.
```

**Comment by AieatAssam** [comment]
My ntp is set to 0.pool.ntp.org in the WiFi settings page and yet during firmware update it seems to be going to pool.ntp.org (note no zero). 
Should time sync failing even be an issue without SSL? Normally it will affect certificate validation but with non ssl this shouldn’t be an issue. 
I’ll still try alternative ntp servers to see if this helps. Will post back results afterwards.

**Comment by AieatAssam** [comment]
I have just changed my ntp to time.google.com in wifi settings, however, it still tries to sync up with pool.ntp.org. I tracked this down to SystemTime.cpp under cores\arduino\system where a hardcoded list of ntp servers is being used. Looking at the code, it seems as if the response from ntp client indicates success when it clearly isn't.
My suspicion is that our corporate network security is causing ntp to time out (it's very time sensitive), so I tethered to my mobile phone. Lo and behold, the update went through and I am now on 1.4
